_drafts/2015-06-29-on-ui-testing.markdown

```diff
@@ -0,0 +1,82 @@
+
+You can mitigate this a little depending on your technology. If you use SpecFlow, then you may have direct access to the code necessary to prime your database. Otherwise, maybe you can create a utility or API endpoints that allow you to populate your data in a more transparent manner. This is the sort of thing that a ReST endpoint can probably do pretty well.
+
+### Mobile
+
+Consensus for UI testing on mobile devices is that it sucks more than that time after the family dinner when our cousin, Toothless Maggie, cornered---...umm... it's pretty bad...
+
+We would love to be proven wrong but to our collective knowledge, there are no decent ways to test a mobile UI in an automated fashion. From what I gather, it's no picnic doing it in a manual fashion. Emulators are laughably bad. And there are more than a few different types and versions of mobile device so you have to use these laughably bad options about a dozen different ways.
+
+### Outsourcing
+
+What about companies that will run through all your test scripts on multiple browsers and multiple devices? You could save some development pain that way. But I personally wouldn't feel comfortable unless the test scripts were _extremely_ prescriptive. And if you're going to that length, you could argue that it's not a large effort to take those prescriptive steps and automate them.
+
+That said, you might get some quick bang for your buck going this route. I've talked to a couple of them and they are always eager to help you. Some of them will even record their test sessions which I would consider a must-have if you decide to use a company for this.
+
+### Tooling
+
+I ain't gonna lie. I like Cucumber and [Capybara](https://github.com/jnicklas/capybara). I've tried [SpecFlow](http://www.specflow.org/) and it's probably as good as you can get in C#, which is decent enough. But it's hard to beat `fill_in 'Email', :with => 'hill@billy.edu'` for conciseness and readability. That said, do **not** underestimate the effort it takes to introduce Ruby to a .NET shop. There is a certain discipline required to maintain your tests and if everyone is scared to dive into your rakefile, you're already mixing stripes with plaid.
+
+We also discussed [Canopy](http://lefthandedgoat.github.io/canopy/) and while none of us has tried it, there was a general appreciation for how it looks. Seems to balance the readability of Capybara with the "it's still .NET" aspect of companies that fear anything non-Microsoft.
+
+Of course, there's Selenium both the IDE and the driver. We mentioned it mostly because you're supposed to.
+
+Some version of Visual Studio also provided support for UI tests, both recorded and coded. The CodedUI tests are supposed to have a pretty nice fluent interface and we generally agreed that coded tests are the way to go instead of recorded ones (as if that were ever in doubt).
+
+Also: [jasmine](http://jasmine.github.io/) and [PhantomJS](http://phantomjs.org/)
```

Comment:
Maybe mention Protractor here: https://angular.github.io/protractor/#/  It is specific to Angular but interesting because it knows to wait until angular is done doing binding things. This avoids the random Sleeps that are common in UI tests.